### PR TITLE
Handle flattened wrapper fields in `fromObject` and `toObject`

### PR DESCRIPTION
**Pull request by dkrol-mix**
This should be what we need. I added all of Ankit's test cases mostly for his info. From his personal tests, 3/4 of them worked. Here, only 1/4 work. Before merging I'll delete the tests that don't pass; the one that passes is the only one that tests the new wrapper handling.

But, if Ankit tells me that the failures are due to me relating them incorrectly, I can fix that and we can have more tests.

### CODE REVIEW

File: tests/api_converters_wrappers.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+tape.test("converters", function(test) {
+
+    protobuf.load("tests/data/convert_wrappers.proto", function(err, root) {
+        if (err)
+            return test.fail(err.message);
+
+        var Message = root.lookup("TestMessage");
+
+        test.test(test.name + " - Message.fromObject - wrappers happy path", function(test) {
```
Comment (dkrol-mix) [inline tests/api_converters_wrappers.js:13]: This one passed.

File: tests/api_converters_wrappers.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            test.same(msg.longField, { low: -1077918720, high: 100689130, unsigned: true }, "should set longField from a number");
+
+            var response = msg.toJSON()
+
+            test.same(response, obj, "original and resultant json representation should be the same")
+
+            test.end();
+        });
+
+        test.test(test.name + " - Message.fromObject wrappers default values", function(test) {
```
Comment (dkrol-mix) [inline tests/api_converters_wrappers.js]: This one had these failures:

```
# converters - Message.fromObject wrappers default values
not ok 77 should set wrappedField from a string
  ---
    operator: deepEqual
    expected: |-
      null
    actual: |-
      { value: '' }
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set wrappedField from a string
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:65:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
not ok 79 should set wrappedId from a number
  ---
    operator: deepEqual
    expected: |-
      null
    actual: |-
      { value: 0 }
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set wrappedId from a number
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:67:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
not ok 81 should set wrappedBool from a boolean
  ---
    operator: deepEqual
    expected: |-
      null
    actual: |-
      { value: false }
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set wrappedBool from a boolean
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:69:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
not ok 83 original and resultant json representation should be the same apart from zero-values
  ---
    operator: deepEqual
    expected: |-
      { id: 1, wrappedField: '', wrappedId: 0, wrappedBool: false, longField: '432456523629741056' }
    actual: |-
      { id: 1, simpleField: '', wrappedField: '', simpleId: 0, wrappedId: 0, simpleBool: false, wrappedBool: false, longField: '432456523629741056' }
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: original and resultant json representation should be the same apart from zero-values
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:74:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
```

File: tests/api_converters_wrappers.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+              wrappedField: "",
+              wrappedId: 0,
+              wrappedBool: false,
+              longField: "432456523629741056"
+            }, "original and resultant json representation should be the same apart from zero-values")
+
+            test.end();
+        });
+
+        test.test(test.name + " - Message.fromObject wrappers null", function(test) {
```
Comment (dkrol-mix) [inline tests/api_converters_wrappers.js]: This one had these failures:

```
# converters - Message.fromObject wrappers null
not ok 87 should set simpleField from a string
  ---
    operator: deepEqual
    expected: null
    actual:   ''
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set simpleField from a string
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:103:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
not ok 89 should set simpleId from a number
  ---
    operator: deepEqual
    expected: null
    actual:   0
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set simpleId from a number
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:105:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
not ok 91 should set simpleBool from a boolean
  ---
    operator: deepEqual
    expected: null
    actual:   false
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set simpleBool from a boolean
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:107:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
```

File: tests/api_converters_wrappers.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+            test.same(response, {
+              id: 1,
+              longField: "432456523629741056"
+            }, "original and resultant json representation should be the same apart from null values")
+
+            test.end();
+        });
+
+        test.test(test.name + " - Message.fromObject wrappers missing", function(test) {
```
Comment (dkrol-mix) [inline tests/api_converters_wrappers.js]: This one had these failures:

```
# converters - Message.fromObject wrappers missing
not ok 98 should set simpleField from a string
  ---
    operator: deepEqual
    expected: null
    actual:   ''
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set simpleField from a string
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:133:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
not ok 100 should set simpleId from a number
  ---
    operator: deepEqual
    expected: null
    actual:   0
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set simpleId from a number
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:135:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
not ok 102 should set simpleBool from a boolean
  ---
    operator: deepEqual
    expected: null
    actual:   false
    at: Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
    stack: |-
      Error: should set simpleBool from a boolean
          at Test.assert [as _assert] (/runtests/node_modules/tape/lib/test.js:225:54)
          at Test.bound [as _assert] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.deepEqual.Test.deepEquals.Test.isEquivalent.Test.same (/runtests/lib/tape-adapter.js:15:10)
          at Test.bound [as same] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.<anonymous> (/runtests/tests/api_converters_wrappers.js:137:18)
          at Test.bound [as _cb] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test.run (/runtests/node_modules/tape/lib/test.js:96:10)
          at Test.bound [as run] (/runtests/node_modules/tape/lib/test.js:77:32)
          at Test._end (/runtests/node_modules/tape/lib/test.js:165:11)
          at Test.bound [as _end] (/runtests/node_modules/tape/lib/test.js:77:32)
  ...
```

File: src/converter.js [hunk 5]
```diff
@@ -32,6 +44,11 @@ function genValuePartial_fromObject(gen, field, fieldIndex, prop) {
                     ("break");
             } gen
             ("}");
+        } else if (WRAPPER_TYPES.indexOf(field.type) !== -1) {
+            gen
+            ("if(typeof d%s===\"object\" && d%s!==null && d%s.value!==undefined)", prop, prop, prop)
+                ("throw TypeError(%j)", field.fullName + ": wrapped value not expected")
+            ("m%s=types[%i].fromObject({value: d%s})", prop, fieldIndex, prop)
```
Comment (dkrol-mix) [inline src/converter.js:51]: Example difference in our output:

```
-                        if (typeof d.creationTimestamp !== "object")
-                            throw TypeError(".mix.data.protobuf.EventHeader.creationTimestamp: object expected");
-                        m.creationTimestamp = $root.google.protobuf.UInt64Value.fromObject(d.creationTimestamp);
+                        if (typeof d.creationTimestamp == "object" && d.creationTimestamp !== null && d.creationTimestamp.value !== undefined)
+                            throw TypeError(".mix.data.protobuf.EventHeader.creationTimestamp: wrapped value not expected");
+                        m.creationTimestamp = $root.google.protobuf.UInt64Value.fromObject({ value: d.creationTimestamp });
```

(See lines 53-55 below to see what generated the previous version)

File: src/converter.js [hunk 6]
```diff
@@ -157,6 +174,10 @@ function genValuePartial_toObject(gen, field, fieldIndex, prop) {
     if (field.resolvedType) {
         if (field.resolvedType instanceof Enum) gen
             ("d%s=o.enums===String?types[%i].values[m%s]:m%s", prop, fieldIndex, prop, prop);
+        else if (WRAPPER_TYPES.indexOf(field.type) !== -1) {
+          gen
+            ("d%s=types[%i].toObject(m%s,o).value", prop, fieldIndex, prop);
+        }
```
Comment (dkrol-mix) [inline src/converter.js:180]: Example difference in our output:

```
-                        d.creationTimestamp = $root.google.protobuf.UInt64Value.toObject(m.creationTimestamp, o);
+                        d.creationTimestamp = $root.google.protobuf.UInt64Value.toObject(m.creationTimestamp, o).value;
```

(See line 182 below to see what generated the previous version)

**Review by alif — APPROVED** [review]
This seems to make sense to me.
I think the tests cover the expectations 👍

**Comment by dkrol-mix** [comment]
To be clear, the first one passes, and the other 3 fail.

**Comment by ankitcha** [comment]
@dkrol-mix - it looks good to me. I think the tests are failing because of #1118 issue which wrongly sets the defaults for missing or null values. 

I would also suggest to only test for json since thats what we are trying to fix with this PR and the internal structures won't be same as expected unless that issue is resolved.

File: tests/api_converters_wrappers.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+            var obj = {
+              id: 1,
+              simpleField: "string value",
+              wrappedField: "string value",
+              simpleId: 1,
+              wrappedId: 1,
+              simpleBool: true,
+              wrappedBool: true,
+              longField: "432456523629741056"
```
Comment (ankitcha) [inline tests/api_converters_wrappers.js]: @dkrol-mix i think we could remove `longField` from these tests. I added to our prototype for validating our specific use-case of 64-bits longs.

File: tests/data/convert_wrappers.proto [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+message TestMessage {
+  uint32 id                                 = 1;
+  string simple_field                       = 2;
+  google.protobuf.StringValue wrapped_field = 3;
+  uint32 simple_id                          = 4;
+  google.protobuf.UInt32Value wrapped_id    = 5;
+  bool simple_bool                          = 6;
+  google.protobuf.BoolValue wrapped_bool    = 7;
+  uint64 long_field                         = 8;
```
Comment (ankitcha) [inline tests/data/convert_wrappers.proto]: @dkrol-mix i think we could remove longField from these tests. I added to our prototype for validating our specific use-case of 64-bits longs.

**Review by ankitcha — APPROVED** [review]
couple of minor comments but otherwise LGTM 👍

**Comment by dkrol-mix** [comment]
See now?

**Review by ankitcha — APPROVED** [review]

**Pull request merged**